### ISSUE TITLE
Expose "increment" parameter to "collate" function in InMemoryDataset

### ISSUE THREAD

**Issue by fierval** [pyg-team/pytorch_geometric#6884]
### 🚀 The feature, motivation and pitch

Currently the `collate` function in `in_memory_dataset.py` around line 96 simply calls the global `collate` and passes `False` as the `increment` argument.

This is not ideal for a case when the graph is merged from a list of independently created subgraphs, where each subgraph is added to the list with its own node/edge indexing:

```python
[
Data(x=[0, 1, 2, 3], edge_index=[[0, 0, 1], [1, 2, 2]), 
Data(x=[0, 1, 2, 3, 4], edge_index=[[0, 1, 1, 2], [3, 4, 2, 0]],...
]
```

In this case collation should increment the indexes when merging the subgraphs, but it's not going to do it since the parameter in charge of that functionality is hard-coded to False.

**Comment by rusty1s** [comment]
It is intended that we do not increment the indices as part of `collate` in `InMemoryDataset`. What's the use-case of wanting to increment here?

**Comment by fierval** [comment]
I think I understand the difference. What you had in mind is collation of several graphs G, such that G1={V, E1}, G2={V, E2},... so you are overlaying different edges on a single set of vertices.

For my case I need to collate a set such that: G1={V1, E1}, G2={V2, E2},... i. e., entirely different graph clusters with completely autonomous edge indices are getting merged together. See the picture, where each colored cluster represents what used to be a separate graph.

![image](https://user-images.githubusercontent.com/987574/224364646-feb80f6f-8ef0-4040-888c-1018f5dc63d2.png)

Once merged, I can split & sample edges through `RandomLinkSplit` for example:

![image](https://user-images.githubusercontent.com/987574/224365823-d139c889-6761-4ce4-a4e5-8ec7d2e3dd2e.png)

**Comment by rusty1s** [comment]
In this case, you can use `Batch.from_data_list`.

**Comment by fierval** [comment]
True. But the design isn't clean then. I prefer to override the base class `collate` but seems unnecessary.
